18.md

```diff
@@ -16,41 +16,34 @@ Payer MAY attach a `payerid` query parameter to LNURL-PAY callback with value se
 
 JSON array may contain one or more records of the following forms:
 
-1. `[ "text/plain", <free-form string> ]`
-2. `[ "application/pubkey", hex(<randomly generated secp256k1 pubkey>) ]`
-3. `[ "application/lnurl-auth", hex(<linkingKey>), hex(sign(utf8ToBytes(<k1>), <linkingPrivKey>)) ]`
-4. `[ "text/identifier", <internet identifier> ]`
-5. `[ "text/email", <email address> ]`
+1. `{ "tag": "name", "value": <free-form string> }`
+2. `{ "tag": "pubkey", "value": hex(<randomly generated secp256k1 pubkey>) }`
+3. `{ "tag": "auth", "value": hex(<linkingKey>), "sig": hex(sign(utf8ToBytes(<k1>), <linkingPrivKey>)) }`
+4. `{ "tag": "identifier", "value": <internet identifier> }`
+5. `{ "tag": "email", "value": <email address> }`
 
-If present, JSON array MAY NOT contain more than one record with the same metadata type.
+## `PayerIDs` record
 
-## Payer identity metadata record
+If `SERVICE` wants to get one or more types of payer identities from `WALLET` then it must alter its JSON response to the first callback to include a `payerIDs` field, as follows:
 
-If `LN SERVICE` wants to get one or more types of payer identities from `LN WALLET` then it MUST put the following record into metadata which contains `application/payer-ids` string as first item further followed by records each indicating a type of identitiy that `LN WALLET` is allowed to provide. 
-
-```
-[
-  ...,
-  [
-    "application/payer-ids", 
-    [ "text/plain" ],
-    [ "application/pubkey" ],
-    [ 
-      "application/lnurl-auth", 
-      isMandatory, // boolean which flags if user MAY include `application/lnurl-auth` record in callback or MUST do that
-      k1 // hex encoded 32 bytes of challenge
-    ],
-    [ "text/identifier" ],
-    [ "text/email" ],
-    ...,
-  ],
-]
+```diff
+ {
+   callback: String,
+   maxSendable: MilliSatoshi,
+   minSendable: MilliSatoshi,
+   metadata: String,
++  payerIDs: {
++    "name": Any,
++    "pubkey": Any,
++    "identifier": Any,
++    "email": Any,
++    "auth": {
++       "mandatory": Boolean,
++       "k1": String // hex encoded 32 bytes of challenge
++    ]
++  },
```

Comment:
How about:

```
payerIDs: {
  "mandatory": JSONArray<String>,
  "optional": JSONArray<String>
}
```

Where the strings in those json arrays could be `name`, `pubkey`, `identifier`, `email`, `auth:random_hex`?

If `Wallet` does not support a mandatory field (e.g `email`), it would reject the lnurl.

Reply:
Good point, indeed. That is flexible enough for other kinds of payerids to be added in the future. Like `home_address` or something like that.

Reply:
But I don't like the `auth:random_hex` part, I'm afraid of string parsing. What about

```
   payerIDs: {
     "name": { "mandatory": Boolean },
     "pubkey": { "mandatory": Boolean },
     "identifier": { "mandatory": Boolean },
     "email": { "mandatory": Boolean },
     "auth": {
        "mandatory": Boolean,
        "k1": String // hex encoded 32 bytes of challenge
     ]
   },
```

Reply:
LGTM 👍